cpp/src/arrow/chunked_array.cc

```diff
@@ -147,13 +148,15 @@ bool ChunkedArray::ApproxEquals(const ChunkedArray& other,
 }
 
 Result<std::shared_ptr<Scalar>> ChunkedArray::GetScalar(int64_t index) const {
-  for (const auto& chunk : chunks_) {
-    if (index < chunk->length()) {
-      return chunk->GetScalar(index);
-    }
-    index -= chunk->length();
+  if (!chunk_resolver_) {
+    chunk_resolver_ = internal::make_unique<internal::ChunkResolver>(chunks_);
```

Comment:
Here as well, there's an atomicity problem. Unfortunately, there are no facilities for atomic access to `unique_ptr`. You'd have to make this a `shared_ptr`; another possibility is to always initialise the `ChunkResolver` in the constructor.

Reply:
Oh, I see. [Immutable objects in Arrow need to be thread-safe](https://arrow.apache.org/docs/cpp/conventions.html?highlight=thread#immutability).

Reply:
For ChunkedArrays with large number of Arrays, there would be a noticeable overhead when creating the offsets for the ChunkResolver. If the application does not accesses the data multiple times (multiple `GetScalar()`), then the overhead may outweigh its potential benefits. On the other hand, using the lazy approach requires synchronization mechanisms.

Reply:
Decided on initializing offsets for `ChunkResolver` during construction as it is the simpler solution and there is no significant performance difference compared to the other solutions.